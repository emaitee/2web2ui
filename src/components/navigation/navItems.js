export default [
  {
    label: 'Dashboard',
    to: '/dashboard',
    icon: 'Home'
  },
  {
    label: 'Reports',
    to: '/reports',
    icon: 'ReportBox',
    children: [
      {
        label: 'Summary',
        to: '/reports/summary'
      }
    ]
  },
  {
    label: 'Templates',
    to: '/templates',
    icon: 'Code'
  },
  {
    label: 'Webhooks',
    to: '/webhooks',
    icon: 'Language'
  },
  {
    label: 'Account',
    to: '/account',
    icon: 'Settings',
    children: [
      {
        label: 'Billing',
        to: '/account/billing'
      },
      {
        label: 'API Keys',
        to: '/account/api-keys'
      },
      {
        label: 'Profile',
        to: '/account/profile'
      }
    ]
  }
];
