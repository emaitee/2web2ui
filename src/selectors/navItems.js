import { createSelector } from 'reselect';
import _ from 'lodash';
import routes from 'src/config/routes';
import navItems from 'src/config/navItems';
import accessConditionState from './accessConditionState';

export const mapNavToRoutes = _.memoize((items) => {
  const routesByPath = _.keyBy(routes, 'path');
  return items.map((item) => {
    item.route = routesByPath[item.to];
    if (item.children) {
      item.children = mapNavToRoutes(item.children);
    }
    return item;
  });
});

export function filterNavByAccess(items, accessConditionState) {
  const firstLevelFiltered = items.filter((item) => {
    const condition = _.get(item, 'route.condition', () => true);
    return condition(accessConditionState);
  });

  const childrenFiltered = firstLevelFiltered.map((item) => {
    if (item.children) {
      item.children = filterNavByAccess(item.children, accessConditionState);
    }
    return item;
  });

  return childrenFiltered.filter((item) => !item.children || item.children.length);
}

export function prepareNavItems(accessConditionState) {
  const mapped = mapNavToRoutes(navItems);
  return filterNavByAccess(mapped, accessConditionState);
}

export default createSelector(
  [accessConditionState],
  prepareNavItems
);
