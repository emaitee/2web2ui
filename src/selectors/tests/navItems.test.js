import { prepareNavItems } from '../navItems';

jest.mock('src/config/routes', () => ([
  {
    path: '/no/children',
    condition: ({ blocked = []}) => !blocked.includes('/no/children')
  },
  {
    path: '/with/children',
    condition: ({ blocked = []}) => !blocked.includes('/with/children')
  },
  {
    path: '/child/a',
    condition: ({ blocked = []}) => !blocked.includes('/child/a')
  },
  {
    path: '/child/b',
    condition: ({ blocked = []}) => !blocked.includes('/child/b')
  },
  {
    path: '/child/c',
    condition: ({ blocked = []}) => !blocked.includes('/child/c')
  }
]));

jest.mock('src/config/navItems', () => ([
  {
    label: 'No Children',
    to: '/no/children'
  },
  {
    label: 'With Children',
    to: '/with/children',
    children: [
      {
        label: 'Child A',
        to: '/child/a'
      },
      {
        label: 'Child B',
        to: '/child/b'
      },
      {
        label: 'Child C',
        to: '/child/c'
      }
    ]
  }
]));

describe('NavItems Selectors', () => {
  let store;

  beforeEach(() => {
    store = {
      blocked: []
    };
  });

  it('should select all nav items if none are blocked', () => {
    const selected = prepareNavItems(store);
    expect(selected).toHaveLength(2);
    expect(selected[1].children).toHaveLength(3);
    expect(selected).toMatchSnapshot();
  });

  it('should remove blocked parents and children', () => {
    store.blocked.push('/no/children');
    store.blocked.push('/child/b');

    const selected = prepareNavItems(store);
    expect(selected).toHaveLength(1);
    expect(selected[0].children).toHaveLength(2);
    expect(selected).toMatchSnapshot();
  });

  it('should remove a parent if all children are blocked', () => {
    store.blocked.push('/child/a');
    store.blocked.push('/child/b');
    store.blocked.push('/child/c');

    const selected = prepareNavItems(store);
    expect(selected).toHaveLength(1);
    expect(selected).toMatchSnapshot();
  });

  it('should remove a parent and all of its children if parent is blocked', () => {
    store.blocked.push('/with/children');

    const selected = prepareNavItems(store);
    expect(selected).toHaveLength(1);
    expect(selected).toMatchSnapshot();
  });

  it('should not mutate items', () => {
    store.blocked.push('/child/a');
    store.blocked.push('/child/b');
    store.blocked.push('/child/c');

    const selected1 = prepareNavItems(store);
    expect(selected1).toHaveLength(1);
    expect(selected1).toMatchSnapshot();

    store.blocked = [];
    const selected2 = prepareNavItems(store);
    expect(selected2).toHaveLength(2);
    expect(selected2[1].children).toHaveLength(3);
    expect(selected2).toMatchSnapshot();
  });

});
