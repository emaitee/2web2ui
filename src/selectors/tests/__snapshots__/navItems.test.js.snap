// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavItems Selectors should filter nav by access of user 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "label": "Summary",
        "route": Object {
          "component": [Function],
          "layout": [Function],
          "path": "/reports/summary",
        },
        "to": "/reports/summary",
      },
      Object {
        "label": "Bounce",
        "route": Object {
          "component": [Function],
          "layout": [Function],
          "path": "/reports/bounce",
        },
        "to": "/reports/bounce",
      },
      Object {
        "label": "Rejections",
        "route": Object {
          "component": [Function],
          "layout": [Function],
          "path": "/reports/rejections",
        },
        "to": "/reports/rejections",
      },
      Object {
        "label": "Accepted",
        "route": Object {
          "component": [Function],
          "layout": [Function],
          "path": "/reports/accepted",
        },
        "to": "/reports/accepted",
      },
      Object {
        "label": "Delayed",
        "route": Object {
          "component": [Function],
          "layout": [Function],
          "path": "/reports/delayed",
        },
        "to": "/reports/delayed",
      },
      Object {
        "label": "Engagement",
        "route": Object {
          "component": [Function],
          "layout": [Function],
          "path": "/reports/engagement",
        },
        "to": "/reports/engagement",
      },
      Object {
        "label": "Message Events",
        "route": Object {
          "component": [Function],
          "layout": [Function],
          "path": "/reports/message-events",
        },
        "to": "/reports/message-events",
      },
    ],
    "icon": "ReportBox",
    "label": "Reports",
    "route": Object {
      "path": "/reports",
      "redirect": "/reports/summary",
    },
    "to": "/reports",
  },
]
`;
